test/shmpreopen.c

```diff
@@ -0,0 +1,234 @@
+
+		// CHECK: foo: [[FOO:.*]]
+		int foo = open(TEST_DIR("/foo"), O_RDONLY | O_DIRECTORY);
+		printf("foo: %d\n", foo);
+		assert(foo != -1);
+		// CHECK: map after foo: [[MAP]]
+		map = po_add(map, "/foo", foo);
+		printf("map after foo: 0x%tx\n", map);
+
+		// CHECK: wibble: [[WIBBLE:.*]]
+		int wibble = po_preopen(map, TEST_DIR("/baz/wibble"));
+		printf("wibble: %d\n", wibble);
+		assert(wibble != -1);
+
+		// CHECK: map after wibble: [[MAP]]
+		map = po_add(map, "/wibble", wibble);
+		printf("map after wibble: 0x%tx\n", map);
+	
+		shmfd=po_create_shmdata(map);
+		
+		i= send_file_descriptor(sockets[1], shmfd);
+		close(sockets[1]);
+	}
+	
+	if(child[1]==0){
+		close(sockets[1]);
+		struct po_map* child_map=NULL;
+		int len,client_shmfd;
+		
+		client_shmfd=recv_file_descriptor(sockets[0]);
+		if(client_shmfd<0){
+			exit(1);
+		}
+		child_map=po_reverse_create_shmdata(client_shmfd);
+		print_map_content(child_map);
+		close(sockets[0]);
+		
+	}	
+	for(k=0;k<2;k++){
+		wait(NULL);
+	}
+
+
+	return 0;
+}
+
+static int
+send_file_descriptor(
+  int socket, /* Socket through which the file descriptor is passed */
+  int fd_to_send) /* File descriptor to be passed, could be another socket */
+{
+ struct msghdr message;
+ struct iovec iov[1];
+ struct cmsghdr *control_message = NULL;
+ char ctrl_buf[CMSG_SPACE(sizeof(int))];
+ char data[1];
+
+ memset(&message, 0, sizeof(struct msghdr));
+ memset(ctrl_buf, 0, CMSG_SPACE(sizeof(int)));
+
+ /* We are passing at least one byte of data so that recvmsg() will not return 0 */
+ data[0] = ' ';
+ iov[0].iov_base = data;
+ iov[0].iov_len = sizeof(data);
+
+ message.msg_name = NULL;
+ message.msg_namelen = 0;
+ message.msg_iov = iov;
+ message.msg_iovlen = 1;
+ message.msg_controllen =  CMSG_SPACE(sizeof(int));
+ message.msg_control = ctrl_buf;
+
+ control_message = CMSG_FIRSTHDR(&message);
+ control_message->cmsg_level = SOL_SOCKET;
+ control_message->cmsg_type = SCM_RIGHTS;
+ control_message->cmsg_len = CMSG_LEN(sizeof(int));
+
+ *((int *) CMSG_DATA(control_message)) = fd_to_send;
+
+ return sendmsg(socket, &message, 0);
+}
+
+// A method for receiving file descriptor
+static int
+recv_file_descriptor(
+  int socket) /* Socket from which the file descriptor is read */
+{
+ int sent_fd;
+ struct msghdr message;
+ struct iovec iov[1];
+ struct cmsghdr *control_message = NULL;
+ char ctrl_buf[CMSG_SPACE(sizeof(int))];
+ char data[1];
+ int res;
+
+ memset(&message, 0, sizeof(struct msghdr));
+ memset(ctrl_buf, 0, CMSG_SPACE(sizeof(int)));
+
+ /* For the dummy data */
+ iov[0].iov_base = data;
+ iov[0].iov_len = sizeof(data);
+
+ message.msg_name = NULL;
+ message.msg_namelen = 0;
+ message.msg_control = ctrl_buf;
+ message.msg_controllen = CMSG_SPACE(sizeof(int));
+ message.msg_iov = iov;
+ message.msg_iovlen = 1;
+
+ if((res = recvmsg(socket, &message, 0)) <= 0)
+  return res;
+
+ /* Iterate through header to find if there is a file descriptor */
+ for(control_message = CMSG_FIRSTHDR(&message);
```

Comment:
This `for` loop could probably be condensed if you renamed `control_message` to something shorter like `header` or even `hdr`.